src/agent.rs

```diff
@@ -296,27 +331,125 @@ impl AgentBuilder {
+    /// let r = agent.get("/my_page").call();
+    /// ```
+    pub fn timeout_write(mut self, timeout: time::Duration) -> Self {
+        self.config.timeout_write = Some(timeout);
+        self
+    }
+
+    /// Timeout for the overall request, including DNS resolution, connection
+    /// time, redirects, and reading the response body. Slow DNS resolution
+    /// may cause a request to exceed the timeout, because the DNS request
+    /// cannot be interrupted with the available APIs.
+    ///
+    /// This takes precedence over `.timeout_read()` and `.timeout_write()`, but
+    /// not `.timeout_connect()`.
+    ///
+    /// ```
+    /// // wait max 1 second for whole request to complete.
+    /// let agent = ureq::builder()
+    ///     .timeout(std::time::Duration::from_secs(1))
+    ///     .build();
+    /// let r = agent.get("/my_page").call();
+    /// ```
+    pub fn timeout(mut self, timeout: time::Duration) -> Self {
+        self.config.timeout = Some(timeout);
+        self
+    }
+
+    /// How many redirects to follow.
+    ///
+    /// Defaults to `5`. Set to `0` to avoid redirects and instead
+    /// get a response object with the 3xx status code.
+    ///
+    /// If the redirect count hits this limit (and it's > 0), TooManyRedirects is returned.
+    ///
+    /// ```
+    /// let r = ureq::builder()
+    ///     .redirects(10)
+    ///     .build()
+    ///     .get("/my_page")
+    ///     .call();
+    /// println!("{:?}", r);
+    /// ```
+    pub fn redirects(mut self, n: u32) -> Self {
+        self.config.redirects = n;
+        self
+    }
+
+    /// Set the TLS client config to use for the connection. See [`ClientConfig`](https://docs.rs/rustls/latest/rustls/struct.ClientConfig.html).
+    ///
+    /// See [`ClientConfig`](https://docs.rs/rustls/latest/rustls/struct.ClientConfig.html).
     ///
     /// Example:
     /// ```
-    /// let proxy = ureq::Proxy::new("user:password@cool.proxy:9090").unwrap();
-    /// let agent = ureq::AgentBuilder::new()
-    ///     .proxy(proxy)
+    /// let tls_config = std::sync::Arc::new(rustls::ClientConfig::new());
+    /// let agent = ureq::builder()
+    ///     .set_tls_config(tls_config.clone())
     ///     .build();
+    /// let req = agent.post("https://cool.server");
     /// ```
-    pub fn proxy(mut self, proxy: Proxy) -> Self {
-        self.proxy = Some(proxy);
+    #[cfg(feature = "tls")]
+    pub fn set_tls_config(mut self, tls_config: Arc<rustls::ClientConfig>) -> Self {
+        self.config.tls_config = Some(TLSClientConfig(tls_config));
         self
     }
 }
 
-pub(crate) fn basic_auth(user: &str, pass: &str) -> String {
-    let safe = match user.find(':') {
-        Some(idx) => &user[..idx],
-        None => user,
-    };
-    base64::encode(&format!("{}:{}", safe, pass))
+#[cfg(feature = "tls")]
+#[derive(Clone)]
+pub(crate) struct TLSClientConfig(pub(crate) Arc<rustls::ClientConfig>);
+
+#[cfg(feature = "tls")]
+impl std::fmt::Debug for TLSClientConfig {
```

Comment:
Nit: elsewhere in the codebase we do `use std::fmt;` and `impl fmt::Debug`; I'd like to do that here to be consistent.

Reply:
Sure. I opted for this because it means conditional on the tls feature. But I don't feel strongly.

Reply:
Oh! I missed that. Is it possible to do:

```rust
#[cfg(feature = "tls")]
{
  use std::fmt;
  impl fmt::Debug for TLSClientConfig {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
  ...
}
```